bot/cogs/print_snippets.py

```diff
@@ -0,0 +1,190 @@
+)
+
+BITBUCKET_RE = re.compile(
+    r'https://bitbucket\.org/(?P<repo>.+?)/src/(?P<branch>.+?)/'
+    + r'(?P<file_path>.+?)#lines-(?P<start_line>\d+)(:(?P<end_line>\d+))?\b'
+)
+
+
+class PrintSnippets(Cog):
+    """
+    Cog that prints out snippets to Discord.
+
+    Matches each message against a regex and prints the contents of all matched snippets.
+    """
+
+    def __init__(self, bot: Bot):
+        """Initializes the cog's bot."""
+        self.bot = bot
+        self.session = aiohttp.ClientSession()
+
+    @Cog.listener()
+    async def on_message(self, message: Message) -> None:
+        """Checks if the message has a snippet link, removes the embed, then sends the snippet contents."""
+        gh_match = GITHUB_RE.search(message.content)
+        gh_gist_match = GITHUB_GIST_RE.search(message.content)
+        gl_match = GITLAB_RE.search(message.content)
+        bb_match = BITBUCKET_RE.search(message.content)
+
+        if (gh_match or gh_gist_match or gl_match or bb_match) and not message.author.bot:
+            message_to_send = ''
+
+            for gh in GITHUB_RE.finditer(message.content):
+                d = gh.groupdict()
+                headers = {'Accept': 'application/vnd.github.v3.raw'}
+                if 'GITHUB_TOKEN' in os.environ:
+                    headers['Authorization'] = f'token {os.environ["GITHUB_TOKEN"]}'
+                file_contents = await fetch_http(
+                    self.session,
+                    f'https://api.github.com/repos/{d["repo"]}'
+                    + f'/contents/{d["file_path"]}?ref={d["branch"]}',
+                    'text',
+                    headers=headers,
+                )
+                message_to_send += await snippet_to_embed(d, file_contents)
+
+            for gh_gist in GITHUB_GIST_RE.finditer(message.content):
+                d = gh_gist.groupdict()
+                gist_json = await fetch_http(
+                    self.session,
+                    f'https://api.github.com/gists/{d["gist_id"]}'
+                    + f'{"/" + d["revision"] if len(d["revision"]) > 0 else ""}',
+                    'json',
+                )
+                for f in gist_json['files']:
+                    if d['file_path'] == f.lower().replace('.', '-'):
+                        d['file_path'] = f
+                        file_contents = await fetch_http(
+                            self.session,
+                            gist_json['files'][f]['raw_url'],
+                            'text',
+                        )
+                        message_to_send += await snippet_to_embed(d, file_contents)
+                        break
+
+            for gl in GITLAB_RE.finditer(message.content):
+                d = gl.groupdict()
+                await orig_to_encode(d)
+                headers = {}
+                if 'GITLAB_TOKEN' in os.environ:
+                    headers['PRIVATE-TOKEN'] = os.environ["GITLAB_TOKEN"]
```

Comment:
Is this optional just for higher rate limits? Does anyone think this key needs to be added to the config?